python/revng/support/artifacts.py

```diff
@@ -0,0 +1,222 @@
+
+    @cached_property
+    def _keys(self) -> List[str]:
+        with tarfile.open(fileobj=BytesIO(self._data)) as tar_file:
+            return tar_file.getnames()
+
+    def __getitem__(self, key: str) -> T:
+        if key not in self._keys:
+            raise KeyError
+        with tarfile.open(fileobj=BytesIO(self._data)) as tar_file:
+            contents = cast(IO[bytes], tar_file.extractfile(key)).read()
+            return self._member_function(contents, self._extracted_mime)
+
+    def __contains__(self, key) -> bool:
+        return key in self._keys
+
+    def __len__(self) -> int:
+        return len(self._keys)
+
+    def __iter__(self):
+        return iter(self._keys)
+
+    def items(self) -> Generator[Tuple[str, T], None, None]:  # type: ignore
+        with tarfile.open(fileobj=BytesIO(self._data)) as tar_file:
+            for member in tar_file.getmembers():
+                contents = cast(IO[bytes], tar_file.extractfile(member)).read()
+                yield (member.name, self._member_function(contents, self._extracted_mime))
+
+    def values(self) -> Generator[T, None, None]:  # type: ignore
+        for _, value in self.items():
+            yield value
+
+
+# Filter function, if it returns 'False' the key will not be processed
+Filter = Callable[[Optional[str]], bool]
+
+
+class _MappedPTMLMixin:
+    def parse(self) -> Dict[str, PTMLDocument]:
+        assert isinstance(self, Mapping)
+        return {k: v.parse() for k, v in self.items()}
+
+    def print(  # noqa: A003
+        self,
+        output: IO[str] = sys.stdout,
+        color: ColorMode = ColorMode.Autodetect,
+        filter_: Optional[Filter] = None,
+    ):
+        assert isinstance(self, Mapping)
+        if filter_ is None:
+            filter_ = lambda x: True  # noqa: E731
+
+        chalk = get_chalk(color)
+        printer: PrinterBackend
+        if color == ColorMode.Off:
+            printer = PlainPrinter(output, indent="  ")
+            key_writer = lambda key: output.write(f"{key}: |-\n  ")  # noqa: E731
+        else:
+            printer = ColorPrinter(output, indent="  ", chalk=chalk)
+            key_color = printer.key_color()
+            key_writer = lambda key: output.write(key_color(f"{key}:") + " |-\n  ")  # noqa: E731
+        for key, value in self.items():
+            if not filter_(key):
+                continue
+
+            key_writer(key)
+            ptml_print_with_printer(value.dump(), printer)
+            output.write("\n")
+
+
+class PTMLTarArtifact(_TarMixin[PTMLArtifact], _MappedPTMLMixin):
+    MIMES = ("text/plain+ptml+tar+gz", "text/x.c+ptml+tar+gz", "text/x.asm+ptml+tar+gz")
+    MEMBER_FUNCTION = PTMLArtifact
+
+
+class TarArtifact(_TarMixin[str]):
+    MIMES = ("text/x.c+tar+gz",)
+    MEMBER_FUNCTION = lambda data, _: data.decode("utf-8")  # noqa; E731
+
+
+class PTMLYamlArtifact(Artifact, _MappedPTMLMixin, Mapping):
```

Comment:
Uppercase `YAML`, not the nicest but still.

Reply:
Done